R/stat_prop.R

```diff
@@ -0,0 +1,239 @@
+#' @importFrom scales percent
+#' @author Joseph Larmarange \email{joseph@@larmarange.net}
+#' @export
+#' @examples
+#'
+#' d <- as.data.frame(Titanic)
+#'
+#' p <- ggplot(d) +
+#'   aes(x = Class, fill = Survived, weight = Freq, by = Class) +
+#'   geom_bar(position = "fill") +
+#'   geom_text(stat = "prop", position = position_fill(.5))
+#' p
+#' p + facet_grid(~ Sex)
+#'
+#' ggplot(d) +
+#'   aes(x = Class, fill = Survived, weight = Freq) +
+#'   geom_bar(position = "dodge") +
+#'   geom_text(
+#'     aes(by = Survived), stat = "prop",
+#'     position = position_dodge(0.9), vjust = "bottom"
+#'  )
+#'
+#' ggplot(d) +
+#'   aes(x = Class, fill = Survived, weight = Freq, by = 1) +
+#'   geom_bar() +
+#'   geom_text(
+#'     aes(label = scales::percent(after_stat(prop), accuracy = 1)),
+#'     stat = "prop",
+#'     position = position_stack(.5)
+#'  )
+#'
+stat_prop <- function(mapping = NULL, data = NULL,
+                       geom = "bar", position = "fill",
+                       ...,
+                       width = NULL,
+                       na.rm = FALSE,
+                       show.legend = NA,
+                       inherit.aes = TRUE) {
+
+  params <- list(
+    na.rm = na.rm,
+    width = width,
+    ...
+  )
+  if (!is.null(params$y)) {
+    stop("stat_prop() must not be used with a y aesthetic.", call. = FALSE)
+  }
+
+  layer(
+    data = data,
+    mapping = mapping,
+    stat = StatProp,
+    geom = geom,
+    position = position,
+    show.legend = show.legend,
+    inherit.aes = inherit.aes,
+    params = params
+  )
+}
+
+#' @rdname stat_prop
+#' @format NULL
+#' @usage NULL
+#' @export
+StatProp <- ggproto("StatProp", Stat,
+  required_aes = c("x", "by"),
+  default_aes = aes(
+    y = stat(count), weight = 1,
+    label = scales::percent(after_stat(prop), accuracy = .1)
+  ),
+
+  setup_params = function(data, params) {
+    if (!is.null(data$y)) {
+      stop("stat_prop() must not be used with a y aesthetic.", call. = FALSE)
+    }
+    # there is an unresolved bug when by is a character vector. To be explored.
+    if (is.character(data$by)) {
+      stop("The by aesthetic should be a factor instead of a character.", call. = FALSE)
+    }
+    params
+  },
+
+  extra_params = c("na.rm"),
+
+  compute_panel = function(self, data, scales, width = NULL) {
+   data$weight <- data$weight %||% rep(1, nrow(data))
+   width <- width %||% (ggplot2::resolution(data$x) * 0.9)
+
+   # sum weights for each combination of by and aesthetics
+   # the use of . allows to consider all aesthetics defined in data
+   panel <- aggregate(weight ~ ., data = data, sum, na.rm = TRUE)
+
+   names(panel)[which(names(panel) == "weight")] <- "count"
+   panel$count[is.na(panel$count)] <- 0
+
+   # compute proportions by by
+   f <- function(x) {sum(abs(x))}
+   panel$prop <- panel$count / ave(panel$count, panel$by, FUN = f)
+   panel$width <- width
+
+   panel
+  }
+)
+
+
+"%||%" <- function(a, b) {
+  if (!is.null(a)) a else b
+}
+
```

Comment:
This is already imported from rlang
```suggestion
"%||%" <- function(a, b) {
  if (!is.null(a)) a else b
}

```

Reply:
ok fixed